src/main.go

```diff
@@ -1,348 +1,170 @@
-			}
-			log.Infoln("initialization time: " + time.Since(p).String())
-		}
-	}()
-	reload <- struct{}{}
-	chainSet.Run()
 
-	apiInstance := api.New(&matrix, conf, configurationFilePath, workers, reload)
-	router := mux.NewRouter()
-	apiInstance.Register(router)
+	brConf, err := brConfig.New(conf.path)
+	if err != nil {
+		printErr("initializing bench-routes configuration", err)
+		return
+	}
 
-	go func() {
-		metrics := sysMetrics.New()
-		type metric struct {
-			cpu    *string
-			memory *sysMetrics.MemoryStats
-			disk   *sysMetrics.DiskStats
-			net    *sysMetrics.NetworkStats
-		}
+	log.Info("msg", "configuration", "out", fmt.Sprintf("%v", brConf.APIs))
 
-		chain := tsdb.NewChain(systemMetricsPath)
-		p := time.Now()
-		chain.Init()
-		log.Infoln("initialized system-metrics...", time.Since(p))
-		chainSet.Register(chain.Name, chain)
+	machineErrCh := make(chan error)
+	machine, err := module.New(module.MachineType, memChainsRef, machineErrCh)
+	if err != nil {
+		printErr("creating machine jobs", err)
+		return
+	}
+	log.Info("msg", "launching machine routine")
+	go machine.Run()
+	if err = machine.Reload(brConf); err != nil {
+		printErr("reloading machine with configuration", err)
+		return
+	}
 
-		for {
-			// collections for cpu, memory and disk run independently and are
-			// time dependent. Hence, running these serially will take more
-			// time than the actual `defaultScrapeTime`. Hence, the
-			// best way is to run them in parallel and get data via channels,
-			// such that defaultScrapeTime >= duration(cpu|memory|disk)
-			// will meet excepted defaultScrapeTime. Anything other
-			// than this will be inaccurate.
-			cpu := make(chan *string)
-			memory := make(chan *sysMetrics.MemoryStats)
-			disk := make(chan *sysMetrics.DiskStats)
-			net := make(chan *sysMetrics.NetworkStats)
+	monitorErrCh := make(chan error)
+	monitor, err := module.New(module.MonitorType, memChainsRef, monitorErrCh)
+	if err != nil {
+		printErr("creating monitor jobs", err)
+		return
+	}
 
-			go metrics.GetTotalCPUUsage(cpu)
-			go metrics.GetVirtualMemoryStats(memory)
-			go metrics.GetDiskIOStats(disk)
-			go metrics.GetNetworkStats(net)
+	log.Info("msg", "launching monitor routine")
+	go monitor.Run()
+	if err = monitor.Reload(brConf); err != nil {
+		printErr("reloading monitor with configuration", err)
+		return
+	}
 
-			data := &metric{
-				cpu:    <-cpu,
-				memory: <-memory,
-				disk:   <-disk,
-				net:    <-net,
+	go func() {
+		// Listen to error channels and exit if any error occurs.
+		// Todo: This can be optimized to accept crashable and non-crashable errors.
+		// Non-crashable errors can be errors from HTTP or Ping request.
+		// Rest all errors (like format error, validation error, etc) are crashable
+		// and hence, the application should stop in those cases.
+		// Right now, we are treating all errors as crashable.
+		for i := 1; i <= 2; i++ {
+			select {
+			case err, ok := <-machineErrCh:
+				if !ok {
```

Comment:
Here we are listening if the `machineErrCh` is closed but we have not closed it anywhere in the code.